packages/server/lib/makeDataContext.ts

```diff
@@ -46,9 +47,20 @@ export function makeDataContext (options: MakeDataContextOptions): DataContext {
       close: browsers.close,
       getBrowsers,
       async ensureAndGetByNameOrPath (nameOrPath: string) {
+        const projectConfig: FullConfig = await (ctx as DataContext).project.getConfig()
+        const declaredBrowsers = projectConfig.browsers?.map<FoundBrowser | undefined>((b) => {
+          if (!b.path) return
```

Comment:
It seems that only browsers injected by the user or plugins have a `path` set so I'm using it as the proxy for custom browsers. Seem reasonable?

Reply:
@ryanjduffy Thank you for opening a change to update improve this check! Considering where this is used, it;'d likely be cheaper and cleaner from a start-up perspective to update `ensureAndGetByNameOrPath` to be a pass through to the imported `ensureAndGetByNameOrPath`

```diff
  getBrowsers,
- async ensureAndGetByNameOrPath (nameOrPath: string) {
-    ...
- },
+ ensureAndGetByNameOrPath
```

and instead update where this is called in the data-context package to supply the list of browsers.
Jumping through the flow:
- we [set the active browser](https://github.com/cypress-io/cypress/blob/develop/packages/data-context/src/data/ProjectLifecycleManager.ts#L267) after we finish loading the config
- this will then call [`ensureAndGetByNameOrPath`](https://github.com/cypress-io/cypress/blob/develop/packages/data-context/src/data/ProjectLifecycleManager.ts#L284) if a browser was provided via the CLI. Seems like we should be able to pull up the machine browsers check and pass it in here.

Reply:
@emilyrohrbough Appreciate the feedback! It's helped me track how this data is used. I've taken a slightly different approach by consolidating both browser lookups (machine and user) into `BrowserDataSource` and then exposed an `allBrowsers` method to fetch both. I like that this approach keeps the "concerns" of the browser config in one spot and makes the diff more clear.

The added bonus of this was that it allowed me to fix `cypress open` to remember the last browser when it was a user browser since that logic can now rely on `allBrowsers`.